### PR TITLE
Leverage the Logstash 8.4.0 DLQ reader to delete consumed segments

### PR DESCRIPTION
**Pull request by andsel**
## Release notes
Expose the `clean_consumed` configuration setting to remove the consumed DLQ segments.

## What does this PR do?

Expose the setting to be used in config files and lock the execution to Logstash `>= 8.4.0` because leverages the API provided by PR https://github.com/elastic/logstash/pull/14188.

## Why is it important/What is the impact to the user?

Permit to the user to automatically clean the DLQ segments once consumed, if the feature is implemented in the target Logstash.

## Checklist

- [x] My code follows the style guidelines of this project
- ~~[ ] I have commented my code, particularly in hard-to-understand areas~~
- ~~[ ] I have made corresponding changes to the documentation~~
- ~~[ ] I have made corresponding change to the default configuration files (and/or docker env variables)~~
- ~~[ ] I have added tests that prove my fix is effective or that my feature works~~

## Author's Checklist

- [x] Build and install the plugin a Logstash and check it removes the segments.

## How to test this PR locally

Testing is very similar to PR https://github.com/elastic/logstash/pull/14255, it needs an upstream pipeline that produces DLQ events to be stored in DLQ segments, and a downstream pipeline with a DLQ input with `clean_consumed` feature enabled. 
Verify that tail segments are removed.

 DLQ feature needs to be enabled in `logstash.yml`.
- prepare the ES index to reject the inserts, create an index and close it:
```
PUT test_index/
POST test_index/_close
```
-  in `logstash.yml` enable DLQ feature:
```
dead_letter_queue.enable: true
```
- run a pipeline that produces events into DLQ:
```
input {
  generator {
    # this is some data
    message => '{"jvm" : {"threads" : {"count" : 49,"peak_count" : 50},"mem" : {"heap_used_percent" : 14,"heap_committed_in_bytes" : 309866496,"heap_max_in_bytes" : 1037959168,"heap_used_in_bytes" : 151686096,"non_heap_used_in_bytes" : 122486176,"non_heap_committed_in_bytes" : 133222400,"pools" : {"survivor" : {"peak_used_in_bytes" : 8912896,"used_in_bytes" : 288776,"peak_max_in_bytes" : 35782656,"max_in_bytes" : 35782656,"committed_in_bytes" : 8912896},"peak_used_in_bytes" : 148656848,"used_in_bytes" : 148656848,"peak_max_in_bytes" : 715849728,"max_in_bytes" : 715849728,"committed_in_bytes" : 229322752},"young" : {"peak_used_in_bytes" : 71630848,"used_in_bytes" : 2740472,"peak_max_in_bytes" : 286326784,"max_in_bytes" : 286326784,"committed_in_bytes" : 71630848}}},"gc" : {"collectors" : {"old" : {"collection_time_in_millis" : 607,"collection_count" : 12},"young" : {"collection_time_in_millis" : 4904,"collection_count" : 1033}}},"uptime_in_millis" : 1809643}'
    codec => json
  }
}

output {
  elasticsearch {
    index => "test_index"
    hosts => "http://localhost:9200"
    user => "elastic"
    password => "changeme"
  }
}
```
- create a downstream pipeline with the `clean_consumed` enabled:
```
input {
  dead_letter_queue {
    path => "/<logstash path.data>/dead_letter_queue/"
    pipeline_id => "dlq_upstream"
    # new feature
    clean_consumed => true
  }
}

output {
  stdout {
    codec => dots
  }
}
```
- verify that the segments are removed once consumed.
```sh
ll /<logstash path.data>/dead_letter_queue/dlq_upstream
```

## Related issues

- 

## Use cases
As user with a DLQ producer and a consumer pipeline,  Logstash as to be able to automatically free the DLQ space used by already consumed events from the consumer side.

### CODE REVIEW

**Comment by andsel** [comment]
The :red_circle:  CI on `ELASTIC_STACK_VERSION=8.x` is expected because the `8.4.0` is not yet published, the important part is the :green_circle:  `SNAPSHOT=true ELASTIC_STACK_VERSION=8.x`.
This is a chicken-egg problem, the new `8.4.0` needs to be released with the DLQ input plugin containing this PR.

File: lib/logstash/inputs/dead_letter_queue.rb [hunk 1]
```diff
@@ -53,10 +58,18 @@ def register
     dlq_path = java.nio.file.Paths.get(File.join(@path, @pipeline_id))
     sincedb_path = @sincedb_path ? java.nio.file.Paths.get(@sincedb_path) : nil
     start_timestamp = @start_timestamp ? org.logstash.Timestamp.new(@start_timestamp) : nil
-    @inner_plugin = org.logstash.input.DeadLetterQueueInputPlugin.new(dlq_path, @commit_offsets, sincedb_path, start_timestamp)
+    @logstash_version = Gem::Version.new(LOGSTASH_CORE_VERSION)
```
Comment (robbavey) [inline lib/logstash/inputs/dead_letter_queue.rb]: Nit: Does this need to be an instance variable?

File: lib/logstash/inputs/dead_letter_queue.rb [hunk 2]
```diff
@@ -53,10 +58,18 @@ def register
...
+    if clean_consumed && !Gem::Requirement.new('>= 8.4.0').satisfied_by?(@logstash_version)
+      raise ConfigurationError.new("clean_consumed can be used only with Logstash version 8.4.0 and above")
+    end
+    if clean_consumed
```
Comment (robbavey) [inline lib/logstash/inputs/dead_letter_queue.rb]: We should log or fail the pipeline if we are implicitly changing a configuration setting, particularly one that is explicitly set

**Review by robbavey — CHANGES_REQUESTED** [review]

File: lib/logstash/inputs/dead_letter_queue.rb (hunk 2 above)
Comment (andsel) [reply lib/logstash/inputs/dead_letter_queue.rb]: Yes we could enforce that the user sets also `commit_offsets`, I thought about the implicit setting once `clean_consumed` to smooth the adoption. But in other plugins we tend to be more explicit, so I'll switch to raising a configuration error.

Comment (andsel) [reply lib/logstash/inputs/dead_letter_queue.rb]: By default `commit_offset` is `true` so if the user doesn't explicitly set to `false`, the adoption is smooth.
Fixed with 114ab84f532174f87c806a46738c622a3cb2dd2e

File: lib/logstash/inputs/dead_letter_queue.rb (hunk 1 above)
Comment (andsel) [reply lib/logstash/inputs/dead_letter_queue.rb]: Fixed by 2a7e21eae2a56b705cc696436c7285bb119f59b3

File: lib/logstash/inputs/dead_letter_queue.rb [hunk 3]
```diff
@@ -53,10 +58,18 @@ def register
...
     sincedb_path = @sincedb_path ? java.nio.file.Paths.get(@sincedb_path) : nil
     start_timestamp = @start_timestamp ? org.logstash.Timestamp.new(@start_timestamp) : nil
-    @inner_plugin = org.logstash.input.DeadLetterQueueInputPlugin.new(dlq_path, @commit_offsets, sincedb_path, start_timestamp)
+    logstash_version = Gem::Version.new(LOGSTASH_CORE_VERSION)
+    if clean_consumed && !Gem::Requirement.new('>= 8.4.0').satisfied_by?(logstash_version)
+      raise ConfigurationError.new("clean_consumed can be used only with Logstash version 8.4.0 and above")
+    end
+    if clean_consumed && !commit_offsets
+      # clean_consumed requires the commit of offset
+      raise ConfigurationError.new("clean_consumed requires that also commit_offsets is enabled to work properly")
```
Comment (robbavey) [inline lib/logstash/inputs/dead_letter_queue.rb]:
```suggestion
      raise ConfigurationError.new("enabling clean_consumed requires commit_offsets to also be enabled")
```

File: spec/unit/inputs/dead_letter_queue_spec.rb [hunk 4]
```diff
@@ -23,6 +26,21 @@
     FileUtils.remove_entry_secure directory
   end
 
+  context "when clean_consumed is enabled" do
+    let(:config) { super().merge!("clean_consumed" => true) }
+
+    it "shouldn't register" do
```
Comment (robbavey) [inline spec/unit/inputs/dead_letter_queue_spec.rb]:
```suggestion
    it "registers successfully" do
```

File: spec/unit/inputs/dead_letter_queue_spec.rb [hunk 5]
```diff
@@ -23,6 +26,21 @@
...
+      expect {subject.register}.to_not raise_error
+    end
+
+    context "and commit_offsets is not" do
+      let(:config) { super().merge!("commit_offsets" => false) }
+      it "shouldn't register" do
```
Comment (robbavey) [inline spec/unit/inputs/dead_letter_queue_spec.rb]: Nit: rspec [style guide](https://github.com/rubocop/rspec-style-guide#should-in-example-docstrings) recommends against using `should` or `shouldn't` in test descriptions

```suggestion
      it "raises a configuration error during register" do
```

File: spec/unit/inputs/dead_letter_queue_spec.rb [hunk 6]
```diff
@@ -23,6 +26,21 @@
...
+        expect {subject.register}.to raise_error
```
Comment (robbavey) [inline spec/unit/inputs/dead_letter_queue_spec.rb]:
```suggestion
        expect {subject.register}.to raise_error(LogStash::ConfigurationError)
```

**Review by robbavey — CHANGES_REQUESTED** [review]
A few nits

**Review by robbavey — APPROVED** [review]
LGTM

**Pull request merged**